### PR TITLE
[expo-sms] Adds Attachment Support To Add Images To SMS Messages

### PR DESCRIPTION
**Pull request by thorbenprimke**
# Why

The Expo SMS package currently does not support image attachments. 

# How

Adds support for both iOS and Android to share an image attachment (png or jpeg) alongside the message. 

***Note***: I would appreciate any feedback on this. If you are open to merging this in, I'll update the documentation as well.

# Test Plan

Added tests to both iOS and Android test suite to validate the functionality. 

||iOS|Android|
|---|---|---|
|Text| Image from iOS | Screenshot_1587636717|
|Text With Image|Image from iOS (1)|Screenshot_1587636660|

### CODE REVIEW

File: apps/test-suite/tests/SMS.ios.js [hunk 1]
```diff
@@ -14,6 +15,40 @@ export function test({ describe, it, expect }) {
...
+          );
+          expect(md5).toBe('1e02045c10b8f1145edc7c8375998f87');
+          await assertExists(true);
+          expect(headers['Content-Type']).toBe('image/png');
+          const contentUri = await FS.getContentUriAsync(localUri);
+          await SMS.sendSMSAsync(['0123456789', '9876543210'], 'test with image', {
+            contentUri,
+            androidType: 'image/png',
+            iOSType: 'public.png',
+            iOSFilename: 'image.png',
```
Comment (EvanBacon) [inline apps/test-suite/tests/SMS.ios.js:47]: This API is pretty disjointed, perhaps we could find a way to remove all android and ios specific fields.

File: packages/expo-sms/ios/EXSMS/EXSMSModule.m [hunk 2]
```diff
@@ -51,6 +52,21 @@ - (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
   messageComposeViewController.messageComposeDelegate = self;
   messageComposeViewController.recipients = addresses;
   messageComposeViewController.body = message;
+
+  if (attachmentData) {
```
Comment (EvanBacon) [inline packages/expo-sms/ios/EXSMS/EXSMSModule.m:56]: I'd rather see an array, something like
```js
for (item of items) {
  const { url, uti = 'public.data', name } = item
}
```

File: packages/expo-sms/ios/EXSMS/EXSMSModule.m [hunk 3]
```diff
@@ -51,6 +52,21 @@ - (void)setModuleRegistry:(UMModuleRegistry *)moduleRegistry
...
+    NSString *contentUri = attachmentData[@"contentUri"];
+    NSString *type = attachmentData[@"iOSType"];
+    NSString *filename = attachmentData[@"iOSFilename"];
+    NSError *error;
+    NSData *attachmentData = [NSData dataWithContentsOfURL:[NSURL URLWithString:contentUri] options:(NSDataReadingOptions)0 error:&error];
+    bool attached = [messageComposeViewController addAttachmentData:attachmentData typeIdentifier:type filename:filename];
+    if (!attached) {
+      reject(@"E_ATTACHMENT", @"Failed to attach", nil);
```
Comment (EvanBacon) [inline packages/expo-sms/ios/EXSMS/EXSMSModule.m:64]:
```suggestion
      reject(@"E_SMS_ATTACHMENT", @"Failed to attach", nil);
```

File: apps/test-suite/tests/SMS.ios.js (hunk 1 above)
Comment (EvanBacon) [reply apps/test-suite/tests/SMS.ios.js:47]: I'd like to see the UTI be aliased to a mime type on iOS so it's universal, and works across SMS and MailComposer.

File: packages/expo-sms/src/__tests__/SMS-test.ts [hunk 4]
```diff
@@ -1,13 +1,34 @@
...
+  expect(ExpoSMS.sendSMSAsync).toHaveBeenLastCalledWith(['0123456789'], 'test', undefined);
+
+  const attachmentData = {
+    contentUri: 'path/myfile.png',
+    iOSFilename: 'myfile.png',
+    iOSType: 'public.png',
+    androidType: 'image/png',
+  } as MMSAttachmentType;
+  await SMS.sendSMSAsync('0123456789', 'test', attachmentData);
+  expect(ExpoSMS.sendSMSAsync).toHaveBeenLastCalledWith(['0123456789'], 'test', attachmentData);
```
Comment (bbarthec) [inline packages/expo-sms/src/__tests__/SMS-test.ts:33]:
```suggestion
  const { ExpoSMS } = NativeModulesProxy;

  const attachmentData = {
    contentUri: 'path/myfile.png',
    iOSFilename: 'myfile.png',
    iOSType: 'public.png',
    androidType: 'image/png',
  } as MMSAttachmentType;
  await SMS.sendSMSAsync('0123456789', 'test', attachmentData);
  expect(ExpoSMS.sendSMSAsync).toHaveBeenLastCalledWith(['0123456789'], 'test', attachmentData);
```

**Review by bbarthec — COMMENTED** [review]
LGTM 👍 
Please add docs and let's make it mergeable! 😊

File: apps/test-suite/tests/SMS.ios.js (hunk 1 above)
Comment (thorbenprimke) [reply apps/test-suite/tests/SMS.ios.js:47]: I think not restricting the mimeType and throwing an error in the event it cannot be matched on the iOS side may be best. 

After your feedback / chatting about it, I tried to add more types (gif, mp3, ...) but it seems like that wouldn't be helpful and would only restrict the usage. 

`isSupportedAttachmentUTI` could be helpful but the Apple's list of mimeType to UTI is pretty exhaustive. I won't be adding this for now unless you feel strongly.

File: packages/expo-sms/ios/EXSMS/EXSMSModule.m (hunk 2 above)
Comment (thorbenprimke) [reply packages/expo-sms/ios/EXSMS/EXSMSModule.m:56]: <a href="https://gitme.me/image?url=https%3A%2F%2Fmedia3.giphy.com%2Fmedia%2FGCvktC0KFy9l6%2Fgiphy.gif&token=cool" data-gitmeme-token="cool"></a>

File: packages/expo-sms/ios/EXSMS/EXSMSModule.m (hunk 3 above)
Comment (thorbenprimke) [reply packages/expo-sms/ios/EXSMS/EXSMSModule.m:64]: done

File: packages/expo-sms/src/__tests__/SMS-test.ts (hunk 4 above)
Comment (thorbenprimke) [reply packages/expo-sms/src/__tests__/SMS-test.ts:33]: 👍

File: apps/test-suite/tests/SMSCommon.js [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+const gifFile = {
+  localUri: FS.documentDirectory + 'image.gif',
+  remoteUri: 'https://upload.wikimedia.org/wikipedia/commons/2/2c/Rotating_earth_%28large%29.gif',
+  md5: '090592ebd01ac1e425b2766989040f80',
+  mimeType: 'image/gif',
+};
+
+const audioFile = {
+  localUri: FS.documentDirectory + 'sound.mp3',
+  remoteUri: 'https://dl.espressif.com/dl/audio/gs-16b-1c-44100hz.mp3',
```
Comment (thorbenprimke) [inline apps/test-suite/tests/SMSCommon.js:45]: @EvanBacon @bbarthec - would one of y'all be able to upload these to the `test-suite-data` S3 bucket?

File: packages/expo-sms/src/SMS.types.ts [hunk 6]
```diff
@@ -1,3 +1,9 @@
 export type SMSResponse = {
   result: 'unknown' | 'sent' | 'cancelled';
 };
+
+export type SMSAttachment = {
+  uri: string;
+  mimeType: string;
```
Comment (thorbenprimke) [inline packages/expo-sms/src/SMS.types.ts:7]: @EvanBacon - I ended up going with `mineType` here to match the parameter naming of `Sharing.shareAsync`'s parameter.

**Comment by thorbenprimke** [comment]
Updated to wrap the `attachments` parameter in an `options` parameter.

File: packages/expo-sms/src/SMS.ts [hunk 7]
```diff
@@ -1,19 +1,41 @@
...
 export { SMSResponse };
 
+function processAttachments(
+  attachments: SMSAttachment | SMSAttachment[] | undefined
+): SMSAttachment[] | null {
+  if (!attachments) {
+    return null;
+  }
+  attachments = Array.isArray(attachments) ? attachments : [attachments];
+  if (Platform.OS === 'android' && attachments.length > 1) {
```
Comment (EvanBacon) [inline packages/expo-sms/src/SMS.ts:16]: If someone forgets, we don't want to warn the end-user
```suggestion
  if (__DEV__ && Platform.OS === 'android' && attachments.length > 1) {
```

File: apps/test-suite/tests/SMSCommon.js (hunk 5 above)
Comment (EvanBacon) [reply apps/test-suite/tests/SMSCommon.js:45]: I can change after merged

**Review by EvanBacon — APPROVED** [review]
Great work on this

**Pull request merged**